File: streaming/bloom_filter.py
"""
bloom_filter.py
----------------
A simple Bloom Filter used to catch duplicate parking-event IDs in the
Kafka -> PySpark streaming pipeline before they get double-counted in
occupancy aggregations.

Why a Bloom Filter here (and not just a Python set or a DB lookup)?
  - Membership checks are O(hash_count), independent of how many events
    have been seen so far.
  - Memory use is fixed up front (a bit array), rather than growing with
    every unique event_id -- important once you're past a few million
    events.
  - The trade-off: it can return a false positive ("might have seen this
    before" when it hasn't), but it will NEVER return a false negative
    ("definitely new" when it's actually a duplicate). That asymmetry is
    fine for duplicate-suppression: false positives just mean we drop an
    occasional new event, but true duplicates are always caught.

Use in Spark: wrap add()/check() as a Python UDF applied per micro-batch
(see streaming/spark_streaming.py for where this plugs in), or maintain
one filter per streaming query and update it via foreachBatch.
"""

import hashlib


class BloomFilter:
    def __init__(self, size: int = 1_000_000, hash_count: int = 5):
        self.size = size
        self.hash_count = hash_count
        self.bit_array = bytearray(size)  # 1 byte per bit, simple & fast for a demo
        self.items_added = 0

    def _hashes(self, item: str):
        positions = []
        for i in range(self.hash_count):
            digest = hashlib.sha256(f"{item}:{i}".encode()).hexdigest()
            positions.append(int(digest, 16) % self.size)
        return positions

    def add(self, item: str) -> None:
        for pos in self._hashes(item):
            self.bit_array[pos] = 1
        self.items_added += 1

    def check(self, item: str) -> bool:
        """Returns True if the item MIGHT have been seen before (possible
        false positive), False if it is DEFINITELY new (no false negatives)."""
        return all(self.bit_array[pos] for pos in self._hashes(item))

    def add_if_new(self, item: str) -> bool:
        """Convenience helper: returns True if the item was new and has now
        been added, False if it looked like a duplicate."""
        if self.check(item):
            return False
        self.add(item)
        return True

    def estimated_false_positive_rate(self) -> float:
        """Standard Bloom Filter FPR estimate: (1 - e^(-k*n/m))^k"""
        k, n, m = self.hash_count, self.items_added, self.size
        return (1 - pow(2.718281828, -k * n / m)) ** k


if __name__ == "__main__":
    bf = BloomFilter(size=100_000, hash_count=5)

    incoming_event_ids = [
        "E00000001", "E00000002", "E00000003",
        "E00000001",  # duplicate
        "E00000004", "E00000002",  # duplicate
    ]

    print("Processing simulated event stream:")
    for eid in incoming_event_ids:
        is_new = bf.add_if_new(eid)
        status = "PROCESS (new)" if is_new else "SKIP (duplicate)"
        print(f"  {eid}: {status}")

    print(f"\nUnique events added: {bf.items_added}")
    print(f"Estimated false-positive rate at this fill level: {bf.estimated_false_positive_rate():.6%}")

File: r-analytics/bloom_filter.R
# bloom_filter.R
# ---------------
# R implementation of a Bloom Filter for duplicate parking-event-id
# detection. This is the R equivalent of streaming/bloom_filter.py, so the
# same "data stream algorithm" experiment (Bloom Filter for membership
# checking) is demonstrated in both languages.
#
# Same trade-off as the Python version: fixed memory footprint (a bit
# vector sized up front), O(hash_count) lookups regardless of how many
# items have been seen, at the cost of a small tunable false-positive
# rate -- but NEVER a false negative. A true duplicate is always caught.
#
# This file only defines the constructor. See bloom_filter_demo.R for a
# runnable demo, and tests/test_bloom_filter.R for unit tests.
#
# Requires: digest (install.packages("digest"))

suppressMessages(library(digest))

BloomFilter <- function(size = 1000000, hash_count = 5) {
  self <- new.env()
  self$size <- size
  self$hash_count <- hash_count
  self$bit_array <- raw(size)       # all-zero bytes = empty filter
  self$items_added <- 0L

  # Derive `hash_count` independent hash positions from one SHA-256 digest
  # per (item, salt) pair -- cheap, and good enough spread for a demo.
  #
  # NOTE: R's strtoi() returns a 32-bit integer, so it silently overflows to
  # NA on hex strings wider than ~8 digits. 7 hex digits (28 bits, max
  # 268,435,455) stays safely inside that range while still giving plenty
  # of spread relative to any realistic filter `size`.
  self$.hashes <- function(item) {
    vapply(seq_len(self$hash_count) - 1L, function(i) {
      h <- digest::digest(paste0(item, ":", i), algo = "sha256", serialize = FALSE)
      (strtoi(substr(h, 1, 7), base = 16L) %% self$size) + 1L  # 1-indexed
    }, integer(1))
  }

  self$add <- function(item) {
    positions <- self$.hashes(item)
    self$bit_array[positions] <- as.raw(1)
    self$items_added <- self$items_added + 1L
    invisible(NULL)
  }

  self$check <- function(item) {
    positions <- self$.hashes(item)
    all(self$bit_array[positions] == as.raw(1))
  }

  # Convenience helper: TRUE if the item was new and has now been added,
  # FALSE if it looked like a duplicate.
  self$add_if_new <- function(item) {
    if (self$check(item)) {
      return(FALSE)
    }
    self$add(item)
    TRUE
  }

  # Standard Bloom Filter false-positive-rate estimate: (1 - e^(-kn/m))^k
  self$false_positive_rate <- function() {
    k <- self$hash_count
    n <- self$items_added
    m <- self$size
    (1 - exp(-k * n / m)) ^ k
  }

  class(self) <- "BloomFilter"
  self
}
